client/js/options.js

```diff
@@ -111,6 +138,84 @@ $("#desktopNotifications").on("change", function() {
 	}
 });
 
+function onPushButton() {
+	pushNotificationsButton.attr("disabled", true);
+
+	navigator.serviceWorker.register("service-worker.js").then((registration) => registration.pushManager.getSubscription().then((existingSubscription) => {
+		if (existingSubscription) {
+			socket.emit("push:unregister", storage.get("token"));
+
+			return existingSubscription.unsubscribe().then((successful) => {
+				if (successful) {
+					alternatePushButton().removeAttr("disabled");
+				}
+			});
+		}
+
+		return registration.pushManager.subscribe({
+			applicationServerKey: urlBase64ToUint8Array(applicationServerKey),
+			userVisibleOnly: true
+		}).then((subscription) => {
+			const rawKey = subscription.getKey ? subscription.getKey("p256dh") : "";
+			const key = rawKey ? window.btoa(String.fromCharCode.apply(null, new Uint8Array(rawKey))) : "";
+			const rawAuthSecret = subscription.getKey ? subscription.getKey("auth") : "";
+			const authSecret = rawAuthSecret ? window.btoa(String.fromCharCode.apply(null, new Uint8Array(rawAuthSecret))) : "";
+
+			socket.emit("push:register", {
+				token: storage.get("token"),
+				endpoint: subscription.endpoint,
+				keys: {
+					p256dh: key,
+					auth: authSecret
+				}
+			});
+
+			alternatePushButton().removeAttr("disabled");
+		});
+	})).catch((err) => {
+		console.error(err);
+		window.alert(err);
+	});
+
+	return false;
+}
+
+if (location.protocol === "https:" || location.hostname === "localhost" || location.hostname === "127.0.0.1") {
+	$("#pushNotificationsHttps").hide();
+
+	if ("serviceWorker" in navigator) {
+		navigator.serviceWorker.register("service-worker.js").then((registration) => {
+			if (!registration.pushManager) {
+				return;
+			}
+
+			return registration.pushManager.getSubscription().then((subscription) => {
+				$("#pushNotificationsUnsupported").hide();
+
+				pushNotificationsButton
+					.removeAttr("disabled")
+					.on("click", onPushButton);
+
+				clientSubscribed = !!subscription;
+
+				if (clientSubscribed) {
+					alternatePushButton();
+				}
+			});
+		}).catch((err) => {
+			$("#pushNotificationsUnsupported p").text(err);
+		});
+	}
+}
+
+function alternatePushButton() {
```

Comment:
So this is to change the text on the button? Feels a little unusual. Why not just have 2 buttons, and only show the appropriate one?